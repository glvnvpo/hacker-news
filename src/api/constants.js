// @flow

type ID = number | string;

const BASE = 'https://hacker-news.firebaseio.com/v0';

export const NEW_STORIES = () => `${BASE}/newstories.json`;
export const ITEM = (id: ID)=> `${BASE}/item/${id}.json`;