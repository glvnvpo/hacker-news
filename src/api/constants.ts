// @flow

import {ID, URL} from '../types';

const BASE = 'https://hacker-news.firebaseio.com/v0';

export const NEW_STORIES = (): URL => `${BASE}/newstories.json`;
export const ITEM = (id: ID): URL => `${BASE}/item/${id}.json`;