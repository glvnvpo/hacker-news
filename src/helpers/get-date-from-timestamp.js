// @flow

export const getDateFromTimestamp = (timestamp: string | number): string => {
	const date = new Date(timestamp*1000);

	const day = date.getDate();
	const month = date.getMonth() + 1;
	const year = date.getFullYear();

	return `${day < 10 ? '0' : ''}${day}.${month < 10 ? '0' : ''}${month}.${year}`;
};