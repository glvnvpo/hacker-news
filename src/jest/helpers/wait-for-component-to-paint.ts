import {act} from 'react-dom/test-utils';

export const waitForComponentToPaint = async () => {
	await act(async () => {
		await new Promise(resolve => setTimeout(resolve, 0));
	});
};