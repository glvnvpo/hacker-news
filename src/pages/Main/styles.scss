@import "~bootstrap/scss/bootstrap";
@import "../../style/variables";

.main {
  display: flex;
  justify-content: center;

  .content {
    width: 50rem;

    .header {
      border-bottom: 2px solid $grey;
      display: flex;
      justify-content: space-between;
      padding-top: 20px;
      padding-bottom: 30px;
      position: sticky;
      top: 0;
      z-index: 1;
    }

    .spinner-border {
      display: flex;
      justify-content: center;
    }

    .cards {
      display: flex;
      flex-direction: column;
      align-items: center;
      margin-bottom: 50px;

      .card {
        width: 100%;

        &:hover {
          border-color: $orange;
          background: $peachy;

          .card-subtitle {
            color: $grey3;
          }
        }
      }
    }
  }
}
