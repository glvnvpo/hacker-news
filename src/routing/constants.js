export const ROOT_PATH                 = '/';
export const MAIN_PAGE_PATH            = '/hacker-news';
export const SINGLE_STORY_PAGE_PATH    = `${MAIN_PAGE_PATH}/:id`;