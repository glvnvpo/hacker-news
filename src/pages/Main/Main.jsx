// @flow

import React, {useEffect, useState} from 'react';
import {useLocation} from 'react-router-dom';
import {useSelector, useDispatch} from 'react-redux';
import {Button, Spinner} from 'react-bootstrap';
import axios from 'axios';
import {isEmpty, isNull} from 'lodash';
import './styles.scss';
import type {Story, ID} from '../../types';
import {setStories} from '../../store/stories';
import {ITEM, NEW_STORIES} from '../../api/constants';
import {MINUTE} from '../../constants/time';
import {MAIN_PAGE_PATH} from '../../routing/constants';
import {StoryCard} from '../../components/StoryCard';

const STORIES_NUMBER = 100;

export const Main = () => {

	const stories = useSelector(state => state.stories.value);
	const dispatch = useDispatch();

	const [isLoading, setLoading] = useState(stories.length <= 0);

	let location = useLocation();

	let timer;

	useEffect(() => {
		loadStoriesEachMinute();
		return () => clearTimeout(timer);
	}, [location?.pathname]);

	const loadStoriesEachMinute = () => {
		loadNewStories();
		timer = setTimeout(loadStoriesEachMinute, MINUTE);
	};

	const loadNewStories = (event = undefined) => {

		if (event) {
			setLoading(true);
		}

		axios<ID[]>(NEW_STORIES())
			.then(async ({data}) => {
				const newestStories = [...data.slice(0, STORIES_NUMBER)];
				let loadedStories = [];
				const promises = newestStories.map(loadOneStory);

				await Promise.allSettled(promises)
					.then(data =>
						loadedStories = data
							.filter(({status}) => status === 'fulfilled')
							.map(({value}) => value)
					)
					.finally(() => {
						dispatch(setStories(loadedStories));
						setLoading(false);
					});
			})
			.catch(() => setLoading(false));
	};

	const loadOneStory = (id: ID) => {
		return new Promise((resolve, reject) => {
			axios<Story>(ITEM(id))
				.then(({data}) => !isNull(data) ? resolve(data) : reject())
				.catch(err => reject(err));
		});
	};

	return (
		<div className='main'>
			<div className='content'>
				<div className='header bg-white'>
					<h4 className='color-grey2'>Latest news</h4>
					<Button onClick={(e) => loadNewStories(e)} variant='outline-success'>Update news</Button>
				</div>

				<div className='cards mt-10'>
					{isLoading ? <Spinner animation='border' variant='secondary' className='mt-20' /> :
						(!isEmpty(stories)) ? stories.map((story: Story) =>
							<StoryCard
								story={story}
								key={story.id}
								asLink
								to={`${MAIN_PAGE_PATH}/${story.id}`}
								extraFieldsToShow={[]}
							/>
						) : <h5 className='mt-20'>No stories found :(</h5>
					}
				</div>

			</div>
		</div>);
};