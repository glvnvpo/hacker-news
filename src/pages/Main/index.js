// @flow

import React, {useState, useEffect} from 'react';
import {Link} from "react-router-dom";
import axios from 'axios';
import {isEmpty} from 'lodash';
import {ITEM, NEW_STORIES} from "../../api/constants";
import {MAIN_PAGE_PATH} from '../../routing/constants';

type Story = {
	id: number | string;
	title: string;
	by: string;
	time: number | string;
	score: number | string;
}

const STORIES_NUMBER = 100;

export const Main = () => {

	const [stories, setStories] = useState([]);

	useEffect(()=>{
		loadNewStories();
	}, []);


	const loadNewStories = () =>{
		axios(NEW_STORIES())
			.then(async ({data}) => {
				const newestStories =[...data.slice(0, STORIES_NUMBER)];
				let arr = [];
				const promises = newestStories.map(loadOneStory);
				await Promise.all(promises)
					.then(p=> arr.push(...p))
					.finally(()=> setStories(arr));
			})
			.catch((err) => console.err(err));
	};

	const loadOneStory = (id: string | number) => {
		return new Promise((resolve, reject)=> {
			axios(ITEM(id))
				.then(({data}) => resolve(data))
				.catch(err => reject(err));
		});
	};

	return (
		<div>
		MAIN PAGE
			{!isEmpty(stories) ? stories.map(({id, title, by, time, score}: Story)=>
				<Link key={id} to={`${MAIN_PAGE_PATH}/${id}`}><li>{title}</li></Link>
			) : 'EMPTY'}
		</div>);
};