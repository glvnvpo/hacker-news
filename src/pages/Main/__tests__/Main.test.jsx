import React from 'react';
import {mount} from 'enzyme';
import axios from 'axios';
import {Provider} from 'react-redux';
import {BrowserRouter} from 'react-router-dom';
import {createStore} from '@reduxjs/toolkit';
import {isNull} from "lodash";
import {Main} from '../Main';
import {NEW_STORIES as mockNEW_STORIES} from "../../../api/constants";
import {waitForComponentToPaint} from "../../../jest/helpers/wait-for-component-to-paint";

let mockLoadNewStories = Promise.resolve({data: [10]});
let mockLoadOneStory = Promise.resolve({
	data: {
		id: 10,
		title: 'someTitle',
		by: 'coolUser',
		time: 1666689031,
		score: 3,
		url: 'https://reactjs.org/',
		descendants: 10
	}
});

jest.mock('axios', () => ({
	__esModule: true,
	...jest.requireActual('axios'),
	default: jest.fn((url) => {
		if (url === mockNEW_STORIES()) {
			return mockLoadNewStories;
		} else return mockLoadOneStory;
	})
}));

jest.mock('lodash', () => ({
	...jest.requireActual('lodash'),
	isNull: jest.fn((value) => value === null)
}));

describe('Main', () => {
	let wrapper;

	let initValue = [{
		id: 20,
		title: 'someTitle',
		by: 'coolUser',
		time: 1666689031,
		score: 4,
		url: 'https://reactjs.org/',
		descendants: 10
	}];
	
	const getWrapper = (initStore = initValue) => {
		const store = createStore(() => ({
			stories: {
				value: initStore
			}
		}));

		return mount(<Provider store={store}>
			<BrowserRouter>
				<Main/>
			</BrowserRouter>
		</Provider>);
	};

	beforeEach(() => {
		jest.clearAllMocks();
	});
	
	it('should render Main page with correct card', () => {
		wrapper = getWrapper();
		expect(wrapper).toMatchSnapshot();
	});

	it('should render Spinner', () => {
		wrapper = getWrapper([]);
		expect(wrapper.find('Spinner').exists()).toBeTruthy();
	});

	it('should call isNull with correct data', async () => {
		mockLoadOneStory = Promise.resolve({
			data: null
		});
		wrapper = getWrapper();
		await waitForComponentToPaint();
		expect(isNull).toHaveBeenCalledWith(null);
	});

	it('should not call isNull', async () => {
		mockLoadOneStory = Promise.reject();
		wrapper = getWrapper();
		await waitForComponentToPaint();
		expect(isNull).not.toHaveBeenCalled();
	});

	it('should call axios twice', () => {
		wrapper = getWrapper();
		waitForComponentToPaint();
		mockLoadNewStories = Promise.reject();
		wrapper.find('.header .btn').simulate('click');
		expect(axios).toHaveBeenCalledTimes(2);
	});
});