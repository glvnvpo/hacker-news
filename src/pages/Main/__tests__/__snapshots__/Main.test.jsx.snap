// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main should render Main page with correct card 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      location={
        Object {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        }
      }
      navigationType="POP"
      navigator={
        Object {
          "action": "POP",
          "createHref": [Function],
          "go": [Function],
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "default",
            "pathname": "/",
            "search": "",
            "state": null,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Main>
        <div
          className="main"
        >
          <div
            className="content"
          >
            <div
              className="header bg-white"
            >
              <h4
                className="color-grey2"
              >
                Latest news
              </h4>
              <Button
                active={false}
                disabled={false}
                onClick={[Function]}
                variant="outline-success"
              >
                <button
                  className="btn btn-outline-success"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Update news
                </button>
              </Button>
            </div>
            <div
              className="cards mt-10"
            >
              <StoryCard
                asLink={true}
                fieldsToShow={
                  Array [
                    "descendants",
                  ]
                }
                key="20"
                story={
                  Object {
                    "by": "coolUser",
                    "descendants": 10,
                    "id": 20,
                    "score": 4,
                    "time": 1666689031,
                    "title": "someTitle",
                    "url": "https://reactjs.org/",
                  }
                }
                to="/hacker-news/20"
              >
                <Card
                  as={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "render": [Function],
                    }
                  }
                  body={false}
                  className="story mt-10 mb-20"
                  to="/hacker-news/20"
                >
                  <Link
                    className="story mt-10 mb-20 card"
                    to="/hacker-news/20"
                  >
                    <a
                      className="story mt-10 mb-20 card"
                      href="/hacker-news/20"
                      onClick={[Function]}
                    >
                      <CardBody>
                        <div
                          className="card-body"
                        >
                          <CardTitle
                            className="color-orange"
                          >
                            <ForwardRef
                              className="color-orange card-title"
                            >
                              <div
                                className="color-orange card-title h5"
                              >
                                someTitle
                              </div>
                            </ForwardRef>
                          </CardTitle>
                          <CardSubtitle
                            className="mb-2 color-grey"
                          >
                            <ForwardRef
                              className="mb-2 color-grey card-subtitle"
                            >
                              <div
                                className="mb-2 color-grey card-subtitle h6"
                              >
                                4
                                 points | 
                                coolUser
                                 | 
                                25.10.2022
                              </div>
                            </ForwardRef>
                          </CardSubtitle>
                          <CardSubtitle
                            className="mb-2 color-grey"
                          >
                            <ForwardRef
                              className="mb-2 color-grey card-subtitle"
                            >
                              <div
                                className="mb-2 color-grey card-subtitle h6"
                              >
                                Comments count: 
                                10
                              </div>
                            </ForwardRef>
                          </CardSubtitle>
                        </div>
                      </CardBody>
                    </a>
                  </Link>
                </Card>
              </StoryCard>
            </div>
          </div>
        </div>
      </Main>
    </Router>
  </BrowserRouter>
</Provider>
`;
